Development/nmos/jwk_utils.cpp

```diff
@@ -277,111 +217,186 @@ namespace nmos
-                if(!rsa)
+                utility::string_t base64_e;
+                if (EVP_PKEY_get_bn_param(pkey.get(), OSSL_PKEY_PARAM_RSA_E, &exponent))
                 {
-                    throw jwk_exception("convert pem to jwk error: failed to load RSA");
+                    base64_e = to_base64url(exponent);
+                    BN_clear_free(exponent);
                 }
 
-                return rsa_to_jwk(rsa, keyid, pubkey_use, alg);
+                // construct jwk
+                return web::json::value_of({
+                    { U("kid"), keyid },
+                    { U("kty"), U("RSA") },
+                    { U("n"), base64_n },
+                    { U("e"), base64_e },
+                    { U("alg"), alg.name },
+                    { U("use"), pubkey_use.name }
+                });
             }
+#endif
+        }
+
+        // extract RSA public key from RSA private key
+        utility::string_t rsa_public_key(const utility::string_t& rsa_private_key)
+        {
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
+            using ssl::experimental::BIO_ptr;
 
-            // convert RSA private key to JSON Web Key
-            web::json::value private_key_to_jwk(const utility::string_t& private_key_, const utility::string_t& keyid, const jwk::public_key_use& pubkey_use, const jwk::algorithm& alg)
+            const std::string private_key_buffer{ utility::us2s(rsa_private_key) };
+            BIO_ptr private_key_bio(BIO_new_mem_buf((void*)private_key_buffer.c_str(), (int)private_key_buffer.length()), &BIO_free);
+            if (!private_key_bio)
             {
-                using ssl::experimental::BIO_ptr;
+                throw jwk_exception("extract public key error: failed to create BIO memory from PEM private key");
+            }
 
-                const std::string buffer{ utility::us2s(private_key_) };
-                BIO_ptr bio(BIO_new_mem_buf((void*)buffer.c_str(), (int)buffer.length()), &BIO_free);
-                EVP_PKEY_ptr private_key(PEM_read_bio_PrivateKey(bio.get(), NULL, NULL, NULL), &EVP_PKEY_free);
+            EVP_PKEY_ptr private_key(PEM_read_bio_PrivateKey(private_key_bio.get(), NULL, NULL, NULL), &EVP_PKEY_free);
 
-                if (private_key)
-                {
-                    RSA_ptr rsa(EVP_PKEY_get1_RSA(private_key.get()), &RSA_free);
-                    if (rsa)
-                    {
-                        return rsa_to_jwk(rsa, keyid, pubkey_use, alg);
-                    }
-                }
-                return{};
+            if (!private_key)
+            {
+                throw jwk_exception("extract public key error: failed to create EVP_PKEY-RSA from BIO private key");
             }
 
-            // find the RSA private key from private key list
-            utility::string_t found_rsa_key(const std::vector<utility::string_t>& private_keys)
+            RSA_ptr rsa(EVP_PKEY_get1_RSA(private_key.get()), &RSA_free);
```

Comment:
In this function is there no difference in control flow, only the difference between these two snippets?

OpenSSL <3.0
```C++
            RSA_ptr rsa(EVP_PKEY_get1_RSA(private_key.get()), &RSA_free);
            if (!rsa)
            {
                throw jwk_exception("extract public key error: failed to load RSA key from private key");
            }
            if (bio && PEM_write_bio_RSA_PUBKEY(bio.get(), rsa.get()))
```
OpenSSL >=3.0
```C++
            if (bio && PEM_write_bio_PUBKEY(bio.get(), private_key.get()))
```

Reply:
Yes, sure, but I thought we wanted the function to be #if around the whole function body rather than just in part of the function.

Reply:
I am happy to change it.

Reply:
Fixed